File: 백준/Gold/16236. 아기 상어/아기 상어.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int N;
    static int[][] map;
    static boolean[][] visited;

    static int[] dr = new int[]{-1, 0, 1, 0};
    static int[] dc = new int[]{0, -1, 0, 1};


    static int size;
    static int time;
    static int cnt;

    static int answer;
    static ArrayList<Node> queue = new ArrayList<>();

    public static class Node implements Comparable<Node> {
        int r;
        int c;


        public Node(int r, int c) {
            this.r = r;
            this.c = c;

        }


        @Override
        public int compareTo(Node o) {
            if(this.r != o.r){
                return this.r - o.r;
            }
            else {
                return this.c - o.c;
            }
        }


    }

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st;


        N = Integer.parseInt(br.readLine());

        map = new int[N][N];
        visited = new boolean[N][N];

        time = 0;
        size = 2;
        cnt = 0;

        Node start = new Node(-1, -1);

        for (int n = 0; n < N; n++) {
            st = new StringTokenizer(br.readLine());
            for (int m = 0; m < N; m++) {
                map[n][m] = Integer.parseInt(st.nextToken());
                if (map[n][m] == 9) {
                    start = new Node(n, m);
                    map[n][m] = 0;
                }
            }
        }

        bfs(start);


    }



    public static void bfs(Node node) {


        queue.add(node);
        boolean flag = false;

        while (!queue.isEmpty()) {

            Collections.sort(queue);
            int qSize = queue.size();

            for (int s = 0; s < qSize; s++) {
                Node now = queue.remove(0);




                int fishSize = map[now.r][now.c];
                if (fishSize > 0 && fishSize < size) {
                    map[now.r][now.c] = 0; // 먹음
                    cnt++; // cnt 증가
                    if (cnt == size) {
                        size++;
                        cnt = 0;
                    }





                    //먹이를 먹었으므로 이동한 것임
                    queue.clear();
                    visited = new boolean[N][N];
                    visited[now.r][now.c] = true;

                    answer = time;

                    flag = true;

                }
                //now에서 이동
                for (int i = 0; i < 4; i++) {
                    int du = now.r + dr[i];
                    int dv = now.c + dc[i];



                    if (du < 0 || dv < 0 || du >= N || dv >= N) continue;

                    if (!visited[du][dv] && map[du][dv] <= size) {
                        queue.add(new Node(du, dv));
                        visited[du][dv] = true;
                    }
                }

                if (flag) {
                    flag = false;
                    break;
                }


            }

            time++;



        }
        System.out.println(answer);

    }
}

File: 백준/Gold/16236. 아기 상어/아기 상어.py
from collections import deque
def find_shark():
    for i in range(n):
        for j in range(n):
            if arr[i][j] == 9:
                arr[i][j] = 0
                return i, j

def bfs(i, j):
    q = deque([(i, j, 1)]) #시작 상어 위치
    visited = [[0]*n for _ in range(n)]
    visited[i][j] = 1
    t = n*n
    ar, ac = n, n
    while q:
        # print(" ================== queue ===================")
        cr, cc, rank = q.popleft()
        # print(cr, cc, rank)
        if rank <= t and arr[cr][cc] != 0 and arr[cr][cc] < size:
            #갱신
            t = rank
            if cr < ar:
                ar, ac = cr, cc
            elif cr==ar and cc<ac:
                ac = cc
            continue
        for di, dj in (-1, 0), (0, -1), (1, 0), (0, 1):
            du = cr+di
            dv = cc+dj

            if du<0 or dv<0 or du>=n or dv>=n:
                continue
            if visited[du][dv] or arr[du][dv] > size:
                continue
            visited[du][dv] = 1
            q.append((du, dv, rank+1))
    return ar, ac, t-1
n = int(input())
arr = [list(map(int, input().split())) for _ in range(n)]
sr, sc = find_shark()
answer = 0
size = 2
cnt  = 0
while 1:
    # print('+=======================')
    # print(sr, sc)
    sr, sc, time = bfs(sr, sc)
    # print(sr, sc, time)
    if sr == n:
        break
    answer += time
    cnt += 1
    arr[sr][sc] = 0

    # print("여기 왜 안돼")
    # for i in range(n):
    #     print(arr[i])
    if cnt == size:
        size+=1
        cnt = 0

print(answer)
